File: tools/android-studio-plugin/src/main/kotlin/com/accessguide/plugin/DeviceManager.kt
package com.accessguide.plugin

import com.intellij.openapi.Disposable
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import kotlinx.coroutines.*
import org.json.JSONArray
import org.json.JSONObject

data class DetectedDevice(val id: String, val platform: String, val name: String)

class DeviceManager(
    private val project: Project,
    private val onDeviceConnected: (DetectedDevice) -> Unit,
    private val onNoDevice: () -> Unit,
) : Disposable {

    private val log = Logger.getInstance(DeviceManager::class.java)
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var flutterProcess: Process? = null
    private var lastDeviceId: String? = null

    fun start() {
        scope.launch { pollLoop() }
    }

    private suspend fun pollLoop() {
        while (scope.isActive) {
            try {
                pollDevices()
            } catch (e: Exception) {
                log.warn("AccessGuide device poll error: ${e.message}")
            }
            delay(5_000)
        }
    }

    private fun pollDevices() {
        val output = ProcessBuilder("flutter", "devices", "--machine")
            .directory(java.io.File(project.basePath ?: return))
            .start()
            .inputStream.bufferedReader().readText()

        val devices = JSONArray(output)
        val physical = (0 until devices.length())
            .map { devices.getJSONObject(it) }
            .filter { !it.optBoolean("isEmulator") && it.optString("id") != "flutter-tester" }

        if (physical.isEmpty()) {
            if (lastDeviceId != null) {
                lastDeviceId = null
                flutterProcess?.destroyForcibly()
                flutterProcess = null
            }
            onNoDevice()
            return
        }

        val device = physical.first()
        val deviceId = device.getString("id")
        val platform = device.optString("targetPlatform", "unknown")
        val name = device.optString("name", deviceId)

        if (deviceId == lastDeviceId && flutterProcess?.isAlive == true) return

        lastDeviceId = deviceId
        launchFlutter(DetectedDevice(deviceId, platform, name))
    }

    private fun launchFlutter(device: DetectedDevice) {
        flutterProcess?.destroyForcibly()

        val projectRoot = project.basePath ?: return
        val args = mutableListOf("flutter", "run", "--device-id", device.id)

        when {
            device.platform.contains("android") -> {
                // Port-forward so app's localhost:9274 → Mac's localhost:9274
                ProcessBuilder("adb", "-s", device.id, "reverse", "tcp:9274", "tcp:9274")
                    .directory(java.io.File(projectRoot))
                    .start()
                args += listOf("--dart-define", "TUTORIAL_HOST=localhost")
            }
            device.platform.contains("ios") -> {
                // iOS on USB can reach the Mac via its LAN IP
                val ip = discoverMacIp()
                args += listOf("--dart-define", "TUTORIAL_HOST=$ip")
            }
        }
        args += listOf("--dart-define", "TUTORIAL_PORT=9274")

        log.info("AccessGuide: launching flutter run on ${device.name}")
        flutterProcess = ProcessBuilder(args)
            .directory(java.io.File(projectRoot))
            .inheritIO()
            .start()

        onDeviceConnected(device)
    }

    private fun discoverMacIp(): String {
        for (iface in listOf("en0", "en1")) {
            try {
                val output = ProcessBuilder("ipconfig", "getifaddr", iface)
                    .start()
                    .inputStream.bufferedReader().readText().trim()
                if (output.matches(Regex("\\d+\\.\\d+\\.\\d+\\.\\d+"))) return output
            } catch (_: Exception) {}
        }
        return "localhost"
    }

    override fun dispose() {
        scope.cancel()
        flutterProcess?.destroyForcibly()
    }
}

File: tools/vscode-extension/src/deviceManager.ts
import { execSync, spawn, ChildProcess } from 'child_process';
import * as vscode from 'vscode';

type FlutterDevice = {
  id: string;
  name: string;
  platform: string;
  isEmulator: boolean;
};

export class DeviceManager {
  private pollTimer: NodeJS.Timeout | null = null;
  private flutterProcess: ChildProcess | null = null;
  private _connectedDevice: FlutterDevice | null = null;
  private onDeviceChange: (device: FlutterDevice | null) => void;

  constructor(onDeviceChange: (device: FlutterDevice | null) => void) {
    this.onDeviceChange = onDeviceChange;
  }

  start(): void {
    this.poll();
    this.pollTimer = setInterval(() => this.poll(), 5000);
  }

  stop(): void {
    if (this.pollTimer) { clearInterval(this.pollTimer); this.pollTimer = null; }
    this.stopFlutter();
  }

  private poll(): void {
    try {
      const raw = execSync('flutter devices --machine 2>/dev/null', { timeout: 8000 }).toString();
      const devices: FlutterDevice[] = JSON.parse(raw);
      // Physical devices only — reject all emulators and simulators.
      const physical = devices.filter(d => !d.isEmulator && d.id !== 'flutter-tester');

      const current = physical[0] ?? null;
      const prevId = this._connectedDevice?.id;

      if (current?.id !== prevId) {
        this._connectedDevice = current;
        this.onDeviceChange(current);
        if (current) {
          this.handleDeviceConnected(current);
        } else {
          this.stopFlutter();
        }
      }
    } catch {
      // flutter not on PATH or timed out — silent fail.
    }
  }

  private handleDeviceConnected(device: FlutterDevice): void {
    // For Android: set up port forwarding.
    if (device.platform.includes('android')) {
      try {
        execSync(`adb -s ${device.id} reverse tcp:9274 tcp:9274`, { timeout: 5000 });
      } catch { /* adb not available */ }
    }

    // Discover the tutorial server host for iOS (Mac's LAN IP).
    const tutorialHost = device.platform.includes('ios')
      ? this.getMacLanIp()
      : 'localhost';

    const workspaceRoot = vscode.workspace.workspaceFolders?.[0]?.uri.fsPath;
    if (!workspaceRoot) return;

    // Don't launch if already running.
    if (this.flutterProcess && !this.flutterProcess.killed) return;

    this.flutterProcess = spawn('flutter', [
      'run',
      '--device-id', device.id,
      '--dart-define', `TUTORIAL_HOST=${tutorialHost}`,
      '--dart-define', `TUTORIAL_PORT=9274`,
    ], {
      cwd: workspaceRoot,
      stdio: 'pipe',
    });

    this.flutterProcess.on('exit', () => { this.flutterProcess = null; });
  }

  private stopFlutter(): void {
    if (this.flutterProcess && !this.flutterProcess.killed) {
      this.flutterProcess.kill();
      this.flutterProcess = null;
    }
  }

  private getMacLanIp(): string {
    try {
      // Try en0 (WiFi), then en1.
      for (const iface of ['en0', 'en1']) {
        const ip = execSync(`ipconfig getifaddr ${iface} 2>/dev/null`).toString().trim();
        if (ip) return ip;
      }
    } catch { /* ignore */ }
    return 'localhost';
  }

  get connectedDevice(): FlutterDevice | null { return this._connectedDevice; }

  dispose(): void { this.stop(); }
}
